www/js/util.js

```diff
@@ -1,44 +1,78 @@
- * modify it under the terms of the GNU General Public License as
- * published by the Free Software Foundation; either version 2 of the
- * License, or (at your option) any later version.
- *
- * Indico is distributed in the hope that it will be useful, but
- * WITHOUT ANY WARRANTY; without even the implied warranty of
- * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
- * General Public License for more details.
- *
- * You should have received a copy of the GNU General Public License
- * along with Indico check-in; if not, see <http://www.gnu.org/licenses/>.
- */
-
-function showAlert(title, text, callback) {
-    navigator.notification.alert(title, callback, text);
-};
-
-function showConfirm(title, text, buttonLabels, callback) {
-    navigator.notification.confirm(text, callback, title, buttonLabels);
-};
-
-String.prototype.hashCode = function () {
-    var hash = 0;
-    if (this.length === 0) {
-        return hash;
-    }
-    for (i = 0; i < this.length; i++) {
-        char = this.charCodeAt(i);
-        hash = ((hash<<5)-hash)+char;
-        hash = hash & hash; // Convert to 32bit integer
-    }
-    return hash;
-};
+// This file is part of Indico.
+// Copyright (C) 2002 - 2021 CERN
+//
+// Indico is free software; you can redistribute it and/or
+// modify it under the terms of the MIT License; see the
+// LICENSE file for more details.
+
+function showAlert(title, text = '', callback = () => {}) {
+  navigator.notification.alert(text, callback, title);
+}
+
+function showConfirm(title, text = '', buttonLabels, callback = () => {}) {
+  navigator.notification.confirm(text, callback, title, buttonLabels);
+}
+
+function generateRandomString(size) {
+  const alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';
+  const randomBytes = window.crypto.getRandomValues(new Uint8Array(size));
+  return Array.from(randomBytes, byte => alphabet[byte % alphabet.length]).join('');
+}
+
+function sha256(message) {
+  const msgUint8 = new TextEncoder().encode(message);
+  return crypto.subtle.digest('SHA-256', msgUint8);
+}
+
+function base64UrlEncode(arrayBuffer) {
+  const str = String.fromCharCode.apply(null, new Uint8Array(arrayBuffer));
+  return btoa(str).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
+}
+
+function buildUrl(baseUrl, params) {
+  const url = new URL(baseUrl);
+  Object.entries(params).forEach(([k, v]) => url.searchParams.append(k, v));
+  return url;
+}
+
+function urlEncodeForm(params) {
+  return Object.keys(params)
+    .map(key => key + '=' + params[key])
+    .join('&');
+}
+
+function getSearchParams(url, searchDivider = '?') {
+  if (url == '' || !url || !url.includes(searchDivider)) return {};
+  const urlParts = url.split(searchDivider);
+
+  const searchParams = {};
+  urlParts[1].split('&').forEach(segment => {
+    [k, v] = segment.split('=');
+    searchParams[k] = v;
+  });
+
+  return searchParams;
+}
+
+function getKey(str) {
```

Comment:
Can't we use some builtin crypto API to get something as simple as an md5 hash here?

Reply:
Kept what was being used, just added the prefix to avoid using the old stored data with the new app.

AFAIK, there is no MD5 built-in in browsers. We could [use any of these](https://developer.mozilla.org/en-US/docs/Web/API/SubtleCrypto/digest#supported_algorithms) but they are more computationally intensive.

If that does not matter,

```javascript
base64UrlEncode(await window.crypto.subtle.digest('SHA-1', (new TextEncoder).encode(str)))
```

might be an option. What do you think?

Reply:
if it was there before i don't mind leaving it. fwiw, i don't think the overhead of sha1 would be noticable here - but if it's an async function then i don't think it's a good idea. (adding way too much complexity for something as simple as hashing)